src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,82 +34,244 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
+     * @param limit
+     * @param allMilestones
+     */
+    private void addMatchingMilestones(List<PickerMilestone> milestonesToBePopulated, int limit,
+                                       List<PickerMilestone> allMilestones) {
+        assert limit > 0;
+        List<PickerMilestone> matchingMilestones = getLimitedMatchingMilestones(limit, allMilestones);
+        milestonesToBePopulated.addAll(matchingMilestones);
     }
 
     /**
-     * Gets the list of milestones that matches the current query
+     * Gets the first {@code limit} milestones from {@code allMilestones} which are not already in
+     * {@code currentMilestones}.
      *
-     * @return
+     * Assumptions: {@code limit} is not negative and {@code allMilestones} should already be ordered
+     * in the desired priority
+     * @param currentMilestones
+     * @param limit
+     * @param allMilestones
      */
-    public List<PickerMilestone> getMatchingMilestonesList() {
-        return this.currentMilestonesList.stream()
-                .filter(milestone -> !milestone.isMatching())
+    private List<PickerMilestone> getLimitedFirstMilestones(List<PickerMilestone> currentMilestones,
+                                                            int limit, List<PickerMilestone> allMilestones) {
+        return allMilestones.stream()
+                .filter(milestone -> !currentMilestones.contains(milestone))
+                .limit(limit)
                 .collect(Collectors.toList());
     }
 
-    private void filterMilestones(String query) {
-        currentMilestonesList
-                .forEach(milestone -> {
-                    boolean matchQuery = Utility.containsIgnoreCase(milestone.getTitle(), query);
-                    if (!milestone.isMatching()) milestone.setMatching(!matchQuery);
-                });
+    /**
+     * Gets a limited number of milestones from {@code allMilestones} matching previous userInputs(s)
+     * which are not already in {@code currentMilestones}.
+     *
+     * Previous userInput(s) refer to userInput as it gets shorter after repetitively removing the last char
+     * e.g. previous userInputs of "milestones" are "milestone", "mileston", "milesto", ..., and they will be
+     * checked for matching milestones in that order
+     *
+     * Assumption: {@code limit} is not negative
+     * @param userInput
+     * @param currentMilestones
+     * @param limit
+     * @param allMilestones
+     */
+    private List<PickerMilestone> getLimitedMatchesFromPreviousInput(String userInput,
+                                                                     List<PickerMilestone> currentMilestones,
+                                                                     int limit, List<PickerMilestone> allMilestones) {
+        List<PickerMilestone> matchesForPreviousInput = new ArrayList<>();
+        String curUserInput = userInput;
+        while (matchesForPreviousInput.size() < limit && curUserInput.length() > 0) {
```

Comment:
Alright.
On Apr 3, 2016 4:44 PM, "Martin" <notifications@github.com> wrote:

> In src/main/java/ui/components/pickers/MilestonePickerState.java
> <https://github.com/HubTurbo/HubTurbo/pull/1439#discussion_r58306586>:
>
> > +     * Previous userInput(s) refer to userInput as it gets shorter after repetitively removing the last char
> > +     * e.g. previous userInputs of "milestones" are "milestone", "mileston", "milesto", ..., and they will be
> > +     * checked for matching milestones in that order
> > +     *
> > +     * Assumption: {@code limit} is not negative
> > +     * @param userInput
> > +     * @param currentMilestones
> > +     * @param limit
> > +     * @param allMilestones
> > +     */
> > +    private List<PickerMilestone> getLimitedMatchesFromPreviousInput(String userInput,
> > +                                                                     List<PickerMilestone> currentMilestones,
> > +                                                                     int limit, List<PickerMilestone> allMilestones) {
> > +        List<PickerMilestone> matchesForPreviousInput = new ArrayList<>();
> > +        String curUserInput = userInput;
> > +        while (matchesForPreviousInput.size() < limit && curUserInput.length() > 0) {
>
> This method already throws an IllegalArgumentException if limit is
> negative (since we can't limit a stream size to a negative number).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/HubTurbo/HubTurbo/pull/1439/files/c936bd2a7838f8c1fefbcc86b28326050c43273e#r58306586>
>